common/ferrostar/src/routing_adapters/osrm/mod.rs

```diff
@@ -83,20 +87,52 @@ impl Route {
             }
         })?;
         if let Some(bbox) = linestring.bounding_rect() {
-            let geometry = linestring
+            let geometry: Vec<GeographicCoordinate> = linestring
                 .coords()
                 .map(|coord| GeographicCoordinate::from(*coord))
                 .collect();
 
+            println!("Route geometry length: {}", geometry.len());
+
             let steps = route
                 .legs
                 .iter()
                 .flat_map(|leg| {
-                    leg.steps
-                        .iter()
-                        .map(|step| RouteStep::from_osrm(step, polyline_precision))
+                    // Converts all single value annotation vectors into a single vector witih a value object.
+                    let annotations = if let Some(leg_annotation) = &leg.annotation {
+                        Some(algorithms::zip_annotations(leg_annotation.clone()))
+                    } else {
+                        None
+                    };
+
+                    // Index for the annotations slice
+                    let mut start_index: usize = 0;
+
+                    return leg.steps.iter().map(move |step| {
+                        let step_geometry =
+                            get_coordinates_from_geometry(&step.geometry, polyline_precision)?;
+
+                        // Slice the annotations for the current step. The annotations array represents segments between coordinates 
+                        // and is thus one element shorter than the geometry array. To account for this, we need to subtract 2 from 
+                        // the length of the geometry array on each step (one to account for the 0-based index, and one to account 
+                        // for the length difference).
+                        let step_index_len = step_geometry.len() - 2 as usize;
+                        let end_index = start_index + step_index_len;
+
+                        let annotation_slice =
+                            get_annotation_slice(start_index, end_index, annotations.clone());
+
+                        start_index = end_index;
```

Comment:
Do steps overlap the start/end coord? If so, this may need to be `end_index - 1`.

Reply:
This was a bit tricky to get right because there are some exceptions. But it's good now and the test `parse_valhalla_asserting_annotation_lengths` helps understand + the snapshot tests. In summary:

1. Annotations should never repeat.
2. Each step has one less annotation than coordinate.
3. The last step never has annotations as it's two of the route's last coordinate (duplicate).

Reply:
This is... a FANTASTIC summary. Could we codify this in a doc comment on the annotations property? (I'm thinking in the _Ferrostar_ generalized route models, not the OSRM ones. That way it's not buried and makes it clear in public docs how any other routing backend integrator could leverage these in a Ferrostar integration as well.)

Reply:
This is now commented in the code that runs the slicing process 👍